nestkernel/node_collection.h

```diff
@@ -88,30 +103,342 @@ class NodeIDTriple
+ *   4. Otherwise, we need to look in the next part.
+ *   5. We first check if there is a next part, if not, we set to end()
+ *   6. Otherwise, we move through remaining parts and use `cumul_abs_size_` to check if we have reached a part
+ *     containing `new_idx`. If so, we also need to check if we ended up in `last_part_` and if so, if before
+ * `last_elem_`.
+ *   7. If we have found a valid element in a new part, we set `part_idx_, element_idx_`, otherwise, we set them to the
+ * end() iterator.
+ *
+ * ### Local iteration
+ *
+ * Rank-local and thread-local iteration work in exactly the same way, just that the period and phase are in one case
+ * given by the ranks and in the other by the threads. Thus, we discuss the algorithms only once.
+ *
+ * - For `period > 1`, the relation of `phase` to position in a given part can differ from part to part. Consider the
+ * following node collection in a simulation with four threads (one MPI process). The table shows the GID of the neuron,
+ * its thread (phase), the part_idx and the element_idx of a iterator pointing to the element. Finally, elements that
+ * belong to thread 1 and 2 respectively, are marked with an asterisk:
+ *
+ *         GID       1  2  3  4  5  6  7  8  9  10  11
+ *         PartIdx   0  0  0  0  0  0  0  0  0   0   0
+ *         ElemIdx   0  1  2  3  4  5  6  7  8   9  10
+ *         Phase     1  2  3  0  1  2  3  0  1   2   3
+ *         -------------------------------------------
+ *         On thr 1  *           *           *
+ *         On thr 2     *           *            *
+ *
+ *     The phase relation here is `phase == element_idx % period + 1`, where `1` is the phase of the node with GID 1.
+ *
+ *     Now consider a new node collection constructed by
+ *
+ *         nc2 = nc[:5] + nc[8:]
+ *
+ *     This yields a node collection with two parts, marked with the vertical line (note that nodes 6 and 7 are not
+ * included). We consider it once it its entirety and once sliced as `nc2[::3]`. The `1stInPt` line marks the elements
+ * indexed by the `first_in_part_` vector.
+ *
+ *         GID            1  2  3  4  5  |  8  9  10  11  12  13  14  15  16
+ *         PartIdx        0  0  0  0  0  |  1  1   1   1   1   1   1   1   1
+ *         ElemIdx        0  1  2  3  4  |  0  1   2   3   4   5   6   7   8
+ *         Phase          1  2  3  0  1  |  0  1   2   3   0   1   2   3   0
+ *         -----------------------------------------------------------------
+ *         All 1stInPt    #              |  #
+ *         All thr 0               *     |  *              *               *
+ *         All thr 1      *           *  |     *               *
+ *         All thr 2         *           |         *               *
+ *         All thr 3            *        |             *               *
+ *         -----------------------------------------------------------------
+ *         [::3] 1stInPt  #              |                     #
```

Comment:
@jstapmanns Thanks for carefully checking the documentation. The figure for the `[::3]` part was plain wrong (but NEST does the right thing). I have now pushed an update, please take another look. In detail:

1. Your calculation is right, the figure was wrong. GID 9 (local index 1) is the first element in `[::3]` in the second part.
2. `[::3]` gives every third element of the full collection, i.e., `[1, 4, 9, 12, 15, 18, 21]`. Because of the gap between 5 and 8, the first element on thread 2 is GID 18. Since we have a stride 3 and a period of 4, only every 12th element on each thread is included.

Reply:
Thank you for the explanation, @heplesser. Now I understand what the implementation should do and I think that it is correct. Thank you!